feat(post): aws transit gateway (#403)
BedrockStreaming/tech.bedrockstreaming.com: _posts/2023-02-24-aws_transit_gateway.md

---
layout: post
title: "Why is the Transit Gateway service not right for us?"
description: "Transit Gateway seems to be the answer to a lot of network problems in AWS. But is it really a good solution?"
author: christianvdz
category:
tags: [on-premise, cloud, aws, network]
color: rgb(251,87,66)
thumbnail: ""
feature-img:
language: en
comments: true

> ---

Reviewer: ```suggestion
excerpt_separator: <!--more-->
---

```

> “Our network architecture became a mess! What can we do about this?”

This is a common question you hear from companies that have grown very quickly.

AWS Transit Gateway seems to be the way out of this. So how do you know if TGW is right for you?

This blog post will introduce how the service works and explain why we chose not to carry on with our migration to AWS Transit Gateway.

> 

Reviewer: ```suggestion
<!--more-->
```

---
layout: post
title: "Why is the Transit Gateway service not right for us?"
description: "Transit Gateway seems to be the answer to a lot of network problems in AWS. But is it really a good solution?"
author: christianvdz
category:
tags: [on-premise, cloud, aws, network]
color: rgb(251,87,66)
thumbnail: ""
feature-img:
language: en
comments: true
excerpt_separator: <!--more-->
---

> Managing the network of many AWS accounts can quickly lead to having a messy network architecture. Transit Gateway service seems to be the way out of this.  

Reviewer 2: "of many AWS accounts", je verrais bien un mot ou deux de plus pour donner le contexte. Peut-être "of many interconnected AWS accounts", ou "of many independant AWS accounts", ou… ?

> So how do you know if TGW is right for you?

Reviewer 2: J'ai l'impression que cette phrase est un peu la seule qui parle à "you" alors que tout le reste de l'article est "we". Ca fait un peu bizarre.

Author: le but de la phrase c'est de remettre en cause l'utilisation de Transit Gateway pour le lecteur

This blog post will introduce how the service works and explain why we chose not to carry on with our migration to AWS Transit Gateway.
<!--more-->
## Transit Gateway’s backstory

Transit Gateway is a network transit hub that connects multiple VPCs and On-Premises sites to allows control traffic between them.  
It was created to provide a new approach of network implementation on AWS and to make network administration smoother.

> VPC peering is a great example of complex network management.

Reviewer 2: why? In one sentence, one example? Why is it complex?
(maybe: IP addresses are shared and must not overlap?)

> ![Network architecture without Transit Gateway](/images/posts/2023-02-24-aws_tgw/tgw.png)

Reviewer 2: Source des illustrations ? C'est toi qui les a faites ? Ou il faut donner la source ?

Author: remplacé par des illustration que je viens de faire

By acting as a “cloud router”, TGW centralizes network connections and takes control of packet forwarding between VPCs.

![Network architecture with Transit Gateway](/images/posts/2023-02-24-aws_tgw/tgw2.png)

> Transit Gateway really does address the complexity and restrictions of VPC peerings.  

Reviewer 2: How does it help? What does it do "better" than the previous approach?

Note: I'm reading this as a non-network-specialist. Maybe I'm not the intented target? It might be useful indicating the intended audience at the beginning of the post.

Author: > "TGW centralizes network connections and takes control of packet forwarding between VPCs."
n'est-ce pas suffisant ?

For now we don’t have any reason not to use it.

## What about TGW at Bedrock?

> We operate more than 20 different AWS accounts for our customers' platforms. Each account has a VPC with at least 3 private and 3 public subnets. We also manage AWS accounts for internal tools like ECR repositories, monitoring tools and shared s3 buckets. We configured Site-to-Site VPNs to all the VPCs in these accounts.

Reviewer 2: Oh, this is useful context ❤️. Maybe put it earlier in the post?

"site to site vpn to all vpcs" -> and from where?

Reviewer 2: A word about regions? One VPC per region, some accounts have 2-3 active regions?

Author: > "Maybe put it earlier in the post?"
will do.
> ""site to site vpn to all vpcs" -> and from where?"
i added "from on premise infrastructure" to the sentence
> "A word about regions? One VPC per region, some accounts have 2-3 active regions?"
not required because we don't have Site-to-Site VPNs to each accounts active regions

From the creation of new AWS accounts to deploying the tenants’ platform, onboarding a new customer requires a lot of work and time.

One of the steps that requires a lot of work is configuring the VPCs Site-to-Site VPN. This is why we were interested in Transit Gateway.

### Proof of concept

We created a production like Proof of Concept infrastructure using three AWS accounts, two different regions, multiple VPCs and a single Site-to-Site VPN.

> #### How we tested TGW ?

Reviewer 2: ```suggestion
#### How did we test TGW ?
```

> - The first thing we tested was splitting routing domains.  

Reviewer 2: why did we test this? what were we expecting to gain from it?

> To do this, we created a Transit Gateway routing table named "dev".  

Reviewer 2: Does the name have any significance for the rest of the post?
If not => remove it (when reading this, one will tend to try to memorize it, thinking it's important later on)

> We associated one of the Transit Gateway attachement to the route table and quickly realized that we wouldn't be able to use attachment route propagation. This means that we have to add routes manually in each route table.

Reviewer 2: "we wouldn't be able to…" -> why? (not supported by aws ?)

Reviewer 2: Not sure if formating this as 3 bullet points is the most lisible way. Maybe 3 paragraphs would be better?

> - The second thing we tested is Transit Gateway peering.  

Reviewer 2: why did we test this? what were we expecting to gain from it?
"link TG from different regions" -> what would we gain from this?

This peering connects allows us to link Transit Gateway's from different regions.  
It is not possible to propagate routes dynamically in the Transit Gateway peering route table.

> - The last thing we tested is migrating from VPC Site-to-Site VPN to TGW VPN.  

Reviewer 2: why did we test this? what were we expecting to gain from it?
does the "minimal down time" apply to the migration? Or to day-to-day operation?

It was important for us to know if we could get a minimal down time on On-Premise to VPC connections.  
This step requires a lot of time because routes have to be deleted and created manually on each side.

Even if we noticed some pain points, tests went well. So we decided to initiate the migration to the Transit Gateway service.

> ### What happened

Reviewer 2: Ce titre pourrait être plus descriptif que "ce qu'il s'est passé". Genre "Les raisons pour lesquelles nous avons fait marche arrière".

Ou tu peux garder une petite section "ce qu'il s'est passé", mais extraire vers une nouvelle section, de plus au niveau, "les raisons de notre retour arrière". Ca ferait une sorte de conclusion à l'article, au lieu que ce soit dans une sous-sous-section ;-)

Everything was okay at first, we successfully migrated two VPC Site-to-Site VPN to our Transit Gateway VPN.

> But then we quickly faced barriers:

Reviewer 2: Reading the post, I'm wondering: why didn't we identify those barriers earlier? Costs, for example, is something that we could/should have calculated even before doing in POC.

Author: We knew about the cost, but we did not notice that we had to pay for each TGW traversal:
on-premise -> bedrock eu-west-1

on premise -> vpn tgw eu-west-3 -> tgw-eu-west-3 -> tgw-eu-west-1 -> vpc bedrock eu-west-1

We pay on VPN traffic and on each TGW traffic

- creating and managing routing domains is possible, but it is impossible to exchanges routes dynamically
- it’s not possible to create routes dynamically in VPC route tables
- data transfer cost is very high (and doubled if a packet goes from on-premise to the transit gateway of one region and then to the transit gateway of the other region)

> - migrating to Transit Gateway requires a planned maintenance because there is a network downtime

Reviewer 2: Ce dernier point, on aurait pu l'accepter, non ? préciser que ce n'est pas un blocker ?

This blog post will introduce how the service works and explain why we chose not to carry on with our migration to AWS Transit Gateway.
<!--more-->
## Transit Gateway’s backstory

Transit Gateway is a network transit hub that connects multiple VPCs and On-Premises sites to allows control traffic between them.  
It was created to provide a new approach of network implementation on AWS and to make network administration smoother.

VPC peering is a point-to-point connections between 2 VPCs.  
It is a great example of complex network management because it adds a new topology to the network architecture.  
On this diagram you can see an example of VPC peering usage. It's not that messy yet but at scale it will be.

![Network architecture without Transit Gateway](/images/posts/2023-02-24-aws_tgw/tgw.png)

By acting as a “cloud router”, TGW centralizes network connections and takes control of packet forwarding between VPCs.  
VPC peerings are not required anymore, we go back to a more simple star network topology.

![Network architecture with Transit Gateway](/images/posts/2023-02-24-aws_tgw/tgw2.png)

Transit Gateway really does address the complexity and restrictions of VPC peerings.  

> For now we don’t have any reason not to use it.

Reviewer 3: I am not a huge fan of double negation. But I have nothing to offer for now :disappointed:

Reviewer 3: ```suggestion
At that point, TGW seems to be the perfect answer for a simpler network configuration/topology.
```

## What about TGW at Bedrock?

> We operate more than 20 different AWS accounts for our customers' platforms. Each account has a VPC with at least 3 private and 3 public subnets. We also manage AWS accounts for internal tools like ECR repositories, monitoring tools and shared s3 buckets. We configured Site-to-Site VPNs from On-Premise infrastructure to all the VPCs in these accounts.

Reviewer 3: ```suggestion
We operate more than 20 different AWS accounts for our customers' platforms. Each account has a VPC with at least 3 private and 3 public subnets. We also manage AWS accounts for internal tools like ECR repositories, monitoring tools and shared s3 buckets. We configured Site-to-Site VPNs from On-Premises infrastructure to all the VPCs in these accounts.
```

From the creation of new AWS accounts to deploying the tenants’ platform, onboarding a new customer requires a lot of work and time.

> One of the steps that requires a lot of work is configuring the VPCs Site-to-Site VPN. This is why we were interested in Transit Gateway.

Reviewer 3: ```suggestion
Configuring VPCs Site-to-Site VPN is one of the steps that requires a lot of work. This is why we were interested in Transit Gateway at first.
```

### Proof of concept

> We created a production like Proof of Concept infrastructure using three AWS accounts, two different regions, multiple VPCs and a single Site-to-Site VPN from TGW to On-Premise firewall.

Reviewer 3: ```suggestion
We created a production like Proof of Concept infrastructure using three AWS accounts, two different regions, multiple VPCs and a single Site-to-Site VPN from TGW to On-Premises firewall.
```

#### How did we test TGW?

We started by trying to split routing domains.  
Centralizing network connections also means (with correct ACLs or Security Groups) that VPCs can reach all other VPCs. We want to control that.  
Transit Gateway attachments read their routes in the TGW route table they are associated to. This is how we manage routing domains.  
We create a Transit Gateway routing table and create routes for target networks.  

> TGW attachments are able to propagate routes in a route table if we want to. But because of routing domains, we can't use that option and we have to add routes manually (attachments only read routes in the route table).

Reviewer 3: >But because of routing domains, we can't use that option and we have to add routes manually (attachments only read routes in the route table).

Not sure to understand this part... What is a `routing domain` in this case?

Then we tested Transit Gateway peering.  
TGW is a regional service, this means that we need to have a TGW for each active AWS region. We use TGW peering to interconnect them.  
We expected to have some way to propagate routes dynamically in the Transit Gateway peering route table. But it is not possible.

The last thing we tested is migrating from VPC Site-to-Site VPN to TGW VPN.  

> Because of the amount of VPC Site-to-Site VPN we have, it was important for us to know if we could get a minimal down time on On-Premise to VPC connections when migrating to the Transit Gateway VPN.  

Reviewer 3: ```suggestion
Because of the amount of VPC Site-to-Site VPN we have, it was important for us to know if we could get a minimal down time on On-Premises to VPC connections when migrating to the Transit Gateway VPN.  
```

This process requires a lot of time because routes have to be deleted and created manually on each side.

Even if we noticed some pain points, tests went well. So we decided to initiate the migration to the Transit Gateway service.

### Why did we choose to rollback?

Everything was okay at first, we successfully migrated two VPC Site-to-Site VPN to our Transit Gateway VPN.

But then previous pain points became barriers:
- creating and managing routing domains is possible, but makes it impossible to use dynamic route propagation

> - there is not option to propagate routes in VPC route table, they all have to be create manually

Reviewer 3: ```suggestion
- there is not option to propagate routes in VPC route table, they all have to be created manually
```

> - data transfer cost is too high (and doubled if packets go through both Transit Gateways)

Reviewer 3: > go through both Transit Gateways  

You said that because we go through TGW on `eu-west-3` and `eu-west-1`, if yes maybe we need to be more explicit on this ?

```suggestion
- data transfer cost is too high (and multiplied by the number of region on which you deployed TGW if your packets go through all these regions)
```

- migrating to Transit Gateway requires a planned maintenance because there is a network downtime

We took some time to talk about what to do next and concluded that migrating to Transit Gateway will just move the complexity of configuring VPC Site-to-Site VPNs to configuring TGW attachments and routes.

> AWS support did not suggest enough solutions to the problems we faced, so we decided to rollback to VPC Site-to-Site VPNs.

Reviewer 3: Don't you also want to point we had configured BGP routing hence transitioning to TGW will not allow us to use BGP anymore ? Or is this out of scope or not relevant ?

Author: When I say "dynamic route propagation" I'm talking about BGP or any equivalent :)
